### DISCUSSION TITLE
WebServer with microdot and utemplate raise Exception when Template is called

### DISCUSSION THREAD

**Discussion by rebe21** [miguelgrinberg/microdot#198] [Q&A]
Hello everyone,
i'm testing the microdot strenghts for a future project with an ES8266 microcontroller, and I'm interested in use utemplate to manage simple html pages (I had previous experience with flask and jinja).

It looks working properly when i call the index page with simple text "Hello World", but
unfortunately i'm having the following exception when I'm try to GET the orders page which requires to be rendered :

`MemoryError: memory allocation failed, allocating 252 bytes` 

Trace is :

Orders - All method
Traceback (most recent call last):
  File "microdot.py", line 1327, in dispatch_request
  File "microdot.py", line 41, in invoke_handler
  File "orders.py", line 15, in index
  File "./utemplate.py", line 34, in __init__
  File "recompile.py", line 22, in load
  File "source.py", line 188, in load
  File "compiled.py", line 14, in load
MemoryError: memory allocation failed, allocating 252 bytes
GET /all 500

The code of orders.py is as follow:

`
      import gc
      from main import app
      from microdot.utemplate import Template

    @app.route('/all', methods=['GET'])
     async def index(req):
        print("Orders - All method")
        name = "mickey"
        orders = ["item1", "item2", "item3"]

        t = Template('orders.html')
        print("Orders - Template created")

        return await t.render_async(name=name, orders=orders)
`

it seems that as soon as the Template is created, the memory exception happens. Anyone knows how to overcome this issue ?
PS. I've already compressed to binary microdot.mpy, utemplate.mpy and other "library" files.

Thanks in advance
Paolo

**Comment by miguelgrinberg** [comment]
How big is your orders.html file?

**Comment by rebe21** [comment]
Hi Miguel, thanks for answering.
The html is pretty small. 18 lines of code. I paste it here below:

>{% args name, orders %}
><!doctype html>
><html>
> <head>
>    <title>Microdot + uTemplate example</title>
>  </head>
>  <body>
>    <h1>List of Orders:</h1>
>    {% if name %}
>    <p>Customer Name: <b>{{ name }}</b>!</p>
>    {% endif %}
>    <ul>
>    {% for order in orders %}
>        <li>{{ order }}</li>
>    {% endfor %}
>    </ul>
>  </body>
></html>

**Reply by rebe21** [reply]
Hi Miguel,
your question brought me on the right track. The issue is due to the html file. I've tried to compact it on a single line deleting all the blank spaces and carriage returns and it worked. 

`{% args name, orders %}<!doctype html><html><body>{% if name %}<p>CN: <b>{{ name }}</b>!</p>{% endif %}<ul>{% for o in orders %}<li>{{ o }}</li>{% endfor %}</ul></body></html>`

So i'm limited to a small html file with this current device (It's a D1 Mini ESp8266 equivalent type with 33Kb of ram)

What if i split the html files in multiple html's files one for each let's say "section" ? 

`{% include 'section1.html' %}
 {% include 'section2.html' %}
 {% include 'section3.html' %}`

Do the framework execute multi requests to load the entire page or it does it build the entire page in memory ?
Would it work ? Otherwise do you have any other suggestions how to optimize it ?

Thanks in advance.
Paolo

**Reply by miguelgrinberg** [reply]
You have two possible optimizations that you can research:

1. You probably noticed that your templates are compiled to `.py` files. You could pre-compile your templates on your desktop/laptop and upload the compiled versions to your device. Then you can initialize Microdot's utemplate extension to not compile templates, just load compiled versions from disk:

    ```python
    from utemplate import compiled
    from microdot.utemplate import Template

    Template.initialize(loader_class=compiled.Loader)
    ```

2. You can switch from `render()` to `generate()` as rendering method. The former generates the entire template in memory before returning it, the latter returns an iterator that renders portions of the template in sequence, without ever having to have the entire template in memory.

    ```python
    return t.generate_async(name=name, orders=orders)
    ```

Let me know if any of this helps!

**Reply by sophiedegran** [reply]
Oo! thanks Miguel I didn't use Template because of that compiled file... Think it will be good to add in documentation ☺

**Reply by hstarmans** [reply]
@miguelgrinberg 
The ```return t.generate_async(name=name, orders=orders``` solved my issues, please use this in your examples (https://github.com/miguelgrinberg/microdot/tree/main/examples/templates/utemplate)  or mention it. It is in the documentation, but I spent quite sometime figuring out how to solve my memory issues.

**Reply by valentingregoire** [reply]
For me, the expanding of the keyword arguments simply didn't work. So in the generated Python template, I would have this:
```python
def render(*args, **kwargs):
    yield """<html><body>
"""
yield str(name)
yield """</body></html>"""
```
My generation would simply be: 
```python
return await Template("index.html").render_async(name="blabla")
```

I simply fixed this by providing my parameters by name in the template, since I will only use these predefined templates:
```python
def render(*args, name):
    yield """<html><body>
"""
yield str(name)
yield """</body></html>"""
```

For loading predefined templates without regenerating them each time, use this:
```python
from microdot.utemplate import Template
from utemplate import compiled

Template.initialize(loader_class=compiled.Loader)
```
